File: python-algorithm/algo_10_greedy/L435-m.py
# -*- coding: utf-8 -*-


# 本题返回的是：要删除的区间的个数
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        sz = len(intervals)
        if sz == 0:
            return 0
        intervals.sort(key = lambda x:x[1])
        ref = intervals[0][1]
        cnt = 0
        for i in range(1, sz):
            if intervals[i][0]<ref:
                cnt += 1
            else:
                ref = intervals[i][1]
        return cnt

File: python-algorithm/algo_10_greedy/L435-m.cpp



//按区间的结尾进行排序，每次选择结尾最小，并且和前一个区间不重叠的区间。

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int eraseOverlapIntervals(vector<vector<int>>& intervals) {
        int sz=intervals.size();
        if(sz == 0) return 0;
        sort(intervals.begin(), intervals.end(), cmp);
        int count=1;
        int end = intervals[0][1];
        for(int i=1; i<sz; ++i){
            if(intervals[i][0]<end)
                continue;
            ++count;
            end = intervals[i][1];
        }
        return sz-count;
    }

    static bool cmp(vector<int>&a, vector<int>&b){
        return a[1] < b[1]; //从小到大排
    }
};